### PR TITLE
Autoencoder plugin implemented.

### PR DESCRIPTION
**Pull request by joeljpoulin**
Several commits are included in this PR, but when PR #140, #146 and #147 will be merged, this current PR will look more clean after rebase.

### CODE REVIEW

File: cortex/built_ins/networks/convnets.py [hunk 1]
```diff
@@ -181,3 +182,51 @@ def forward(self, x, nonlinearity=None, **nonlinearity_args):
...
+        return self.nets.encoder(inputs, **kwargs)
+
+    def visualize(self, inputs, targets):
+        Z = self.encode(inputs)
+        if targets is not None:
+            targets = targets.data
+        self.add_scatter(Z.data, labels=targets, name='latent values')
+
+
+class ImageDecoder(ModelPlugin):
```
Comment (rdevon) [inline cortex/built_ins/networks/convnets.py]: I agree these should go in a separate file, but not here. These files are for built in networks. What you copied here are "models". How about moving these two models into a separate file in built_ins/models?

**Comment by rdevon** [comment]
Remove the rst files please.

File: cortex/built_ins/models/ae.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        self.losses.ae = r_loss
+
+    def visualize(self, inputs, targets):
+        ae = self.nets.ae
+        outputs = ae(inputs)
+        self.add_image(outputs, name='reconstruction')
+        self.add_image(inputs, name='ground truth')
+
+
+if __name__ == '__main__':
```
Comment (rdevon) [inline cortex/built_ins/models/ae.py:43]: So this shouldn't be in a built_in model. Rather, add "ae" to `__all__` in `__init__.py`. This functionality is for using models *external* to cortex.

**Comment by joeljpoulin** [comment]
@rdevon We will need to keep them, as they are not all auto-generated, they are considered as source files for documentation and then needs to stay in the repository, because multiple contributors can make changes to the same .rst pages and that GitHub Pages integration needs documentation source files to build the HTML files and deploy it on github.io.

**Comment by rdevon** [comment]
@joeljpoulin  Let's move this PR to dev now that mine is merged in #140 .

File: cortex/built_ins/models/ae.py (hunk 2 above)
Comment (joeljpoulin) [reply cortex/built_ins/models/ae.py]: The auto-encoder has been added as a demo (6d6933e).

File: cortex/built_ins/networks/convnets.py (hunk 1 above)
Comment (joeljpoulin) [reply cortex/built_ins/networks/convnets.py]: They have been moved to a file called image_coders.py inside /cortex/built-ins/models (6d6933e).

**Comment by joeljpoulin** [comment]
@rdevon The auto-encoder has been moved as a demo, and changes requested have been made. The autoencoder is still using Cortex built-ins for this PR.

File: cortex/built_ins/models/ae.py (hunk 2 above)
Comment (rdevon) [reply cortex/built_ins/models/ae.py:43]: This is fine, though having the autoencoder as a built-in is also desirable (one serves as a demo, the other serves to enhance cortex). I'm fine with either demo, built in, or both in this PR

**Pull request merged**